src/v/storage/disk_log_impl.cc

```diff
@@ -2511,7 +2511,12 @@ ss::future<>
 disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
     return ss::do_until(
       [this, cfg] {
+          // base_offset check is for the case of an empty segment
+          // (where dirty = base - 1). We don't want to remove it because
+          // batches may be concurrently appended to it and we should keep them.
```

Comment:
yikes. we probably should have had a lot more concurrency limitations in disk_log_impl. another thing that @andrwng's work on local storage v2 would address.